lgc/patch/Patch.cpp

```diff
@@ -180,8 +184,57 @@ void Patch::addPasses(PipelineState *pipelineState, lgc::PassManager &passMgr, b
     LgcContext::createAndAddStartStopTimer(passMgr, patchTimer, true);
   }
 
-  // NOTE: The new pass manager is not fully implemented yet. We should add all
-  // the other patching passes here.
+  // Patch buffer operations (must be after optimizations)
+  passMgr.addPass(createModuleToFunctionPassAdaptor(PatchBufferOp()));
+  passMgr.addPass(createModuleToFunctionPassAdaptor(InstCombinePass(2)));
+
+  // Fully prepare the pipeline ABI (must be after optimizations)
+  passMgr.addPass(PatchPreparePipelineAbi(/* onlySetCallingConvs = */ false));
+
+  const bool canUseNgg = pipelineState->isGraphics() && pipelineState->getTargetInfo().getGfxIpVersion().major == 10 &&
+                         (pipelineState->getOptions().nggFlags & NggFlagDisable) == 0;
+  if (canUseNgg) {
+    // Stop timer for patching passes and restart timer for optimization passes.
+    if (patchTimer) {
+      LgcContext::createAndAddStartStopTimer(passMgr, patchTimer, false);
+      LgcContext::createAndAddStartStopTimer(passMgr, optTimer, true);
+    }
+
+    // Extra optimizations after NGG primitive shader creation
+    passMgr.addPass(AlwaysInlinerPass());
+    passMgr.addPass(GlobalDCEPass());
+    passMgr.addPass(createModuleToFunctionPassAdaptor(PromotePass()));
+    passMgr.addPass(createModuleToFunctionPassAdaptor(ADCEPass()));
+    passMgr.addPass(createModuleToFunctionPassAdaptor(InstCombinePass()));
+    passMgr.addPass(createModuleToFunctionPassAdaptor(SimplifyCFGPass()));
```

Comment:
According to the [llvm docs](https://llvm.org/docs/NewPassManager.html#adding-passes-to-a-pass-manager):
> Generally you want to group CGSCC/function/loop passes together in a pass manager, as opposed to adding adaptors for each pass to the containing upper level pass manager

because it’s beneficial for cache locality.
So I guess it makes sense to group the passes.

```cpp
FunctionPassManager FPM;
FPM.addPass(FunctionPass1());
FPM.addPass(FunctionPass2());

MPM.addPass(createModuleToFunctionPassAdaptor(std::move(FPM)));
```